xdsl/ir/core.py

```diff
@@ -456,6 +462,96 @@ def print_parameter(self, printer: Printer) -> None:
         """Print the attribute parameter."""
 
 
+class StrEnum(str, Enum):
+    """
+    Homemade StrEnum. StrEnum is standard in Python>=3.11.
+    """
+
+    @staticmethod
+    def _generate_next_value_(
+        name: str, start: int, count: int, last_values: list[Any]
+    ):
+        return name.lower()
+
+    def __str__(self) -> str:
+        return self.value
+
+
+EnumType = TypeVar("EnumType", bound=StrEnum)
+
+
+class EnumAttribute(Data[EnumType]):
+    """
+    Core helper for Enum Attributes. Takes a StrEnum type parameter, and defines
+    parsing/printing automatically from its values, restricted to be parsable as
+    identifiers.
+
+    example:
+    ```python
+    class MyEnum(StrEnum):
+        First = auto()
+        Second = auto()
+
+    class MyEnumAttribute(EnumAttribute[MyEnum], OpaqueSyntaxAttribute):
+        name = "example.my_enum"
+    ```
+    Is enough to use this attribute, having a textual representation
+    of `example<my_enum first>` and ``example<my_enum second>``
+
+    """
+
+    enum_type: ClassVar[type[StrEnum]]
+
+    def __init_subclass__(cls) -> None:
+        """
+        This hook first checks two constraints, enforced to keep the implementation
+        reasonable, until more complex usecases appears. It then store the Enum type
+        used by the subclass to use in parsing/printing.
+
+        The constraints are:
+
+        - Only direct, specialized inheritance is allowed. That is, using a subclass
+        of EnumAttribute as a base class is *not supported*.
+          This simplifies type-hacking code and I don't see it being too restrictive
+          anytime soon.
+        - The StrEnum values must all be parsable as identifier. This is to keep the
+        parsing code simple and efficient. This restriction is easier to lift, but I
+        didn't yet meet an example use case where it matters, so I'm keeping it simple.
+        """
+        orig_bases = getattr(cls, "__orig_bases__")
+        enumattr = next(b for b in orig_bases if get_origin(b) is EnumAttribute)
+        enum_type = get_args(enumattr)[0]
+        if isinstance(enum_type, TypeVar):
+            raise TypeError("Only direct inheritance from EnumAttribute is allowed.")
+
+        for v in enum_type:
+            if lexer.Lexer.bare_identifier_suffix_regex.fullmatch(v) is None:
+                raise ValueError(
+                    "All StrEnum values of an EnumAttribute must be parsable as an identifer."
+                )
+
+        cls.enum_type = enum_type
+
+    @final
+    def print_parameter(self, printer: Printer) -> None:
+        printer.print(" ", self.data.value)
+
+    @final
+    @classmethod
+    def parse_parameter(cls, parser: AttrParser) -> EnumType:
+        enum_type = cls.enum_type
+
+        val = parser.parse_identifier()
+        if val in enum_type.__members__.values():
+            return cast(EnumType, enum_type(val))
```

Comment:
super nit: it feels a little weird to have the success condition not be at the bottom of the function. If you feel like doing it, I feel that it would be a little clearer to invert the if condition and raise the error in the if block. But not a big deal

Reply:
Agreed and I don't think it's a big deal, will do 🙂